### PR TITLE
More opt

### PR DESCRIPTION
**Pull request by Therzok**
cc @DavidKarlas

### CODE REVIEW

**Comment by Therzok** [comment]
These are the [stats](https://github.com/Therzok/GtkBench/commit/e64562cece482e5d1d08ed793c94b0585664b12e) for https://github.com/mono/gtk-sharp/pull/153/commits/7fedfdd2b622ffc700e0f83ab8923a78e4a2a54b

File: glib/Object.cs [hunk 1]
```diff
@@ -390,12 +390,17 @@ protected Object (GType gtype)
 		[DllImport("glibsharpglue-2", CallingConvention=CallingConvention.Cdecl)]
 		static extern IntPtr gtksharp_object_newv (IntPtr gtype, int n_params, IntPtr[] names, GLib.Value[] vals);
 
-		protected virtual void CreateNativeObject (string[] names, GLib.Value[] vals)
+		protected void CreateNativeObject (string[] names, GLib.Value[] vals)
```
Comment (alanmcgovern) [inline glib/Object.cs]: is this an api breaking change?

File: glib/Marshaller.cs [hunk 2]
```diff
@@ -23,12 +23,11 @@
 
 namespace GLib {
 	using System;
+	using System.Linq;
 	using System.Runtime.InteropServices;
 	
-	public class Marshaller {
+	public static class Marshaller {
```
Comment (alanmcgovern) [inline glib/Marshaller.cs]: this is api breaking too, if we care

File: cairo/Context.cs [hunk 3]
```diff
@@ -163,7 +163,8 @@ static Context ()
 			// except in the case of Win64 where sizeof(long)
 			// is 32 bits
 			//
-			int ptr_size = Marshal.SizeOf (typeof (IntPtr));
+			int ptr_size;
+			unsafe { ptr_size = sizeof (IntPtr); }
```
Comment (alanmcgovern) [inline cairo/Context.cs]: `IntPtr.Size`?

File: gtk/TreeStore.custom [hunk 4]
```diff
@@ -475,7 +475,7 @@
 			vals [2] = new GLib.Value (iter);
 			inst_and_params.Append (vals [2]);
 			int cnt = IterNChildren (iter);
-			IntPtr new_order_ptr = Marshal.AllocHGlobal (Marshal.SizeOf (typeof (int)) * cnt);
+			IntPtr new_order_ptr = Marshal.AllocHGlobal (sizeof (int) * cnt);
```
Comment (alanmcgovern) [inline gtk/TreeStore.custom]: for all of these changes,

```
Marshal.SizeOf returns the size after the type has been marshaled, whereas sizeof returns the size as it has been allocated by the common language runtime, including any padding
```
https://msdn.microsoft.com/en-us/library/eahchzkf.aspx

Might need to undo this on the grounds of correctness.

File: glib/Object.cs (hunk 1 above)
Comment (Therzok) [inline glib/Object.cs]: Kinda, but nobody really uses that API except generated classes.

File: glib/Marshaller.cs (hunk 2 above)
Comment (Therzok) [inline glib/Marshaller.cs]: ABI yes, API no. The constructor is private.

File: cairo/Context.cs (hunk 3 above)
Comment (Therzok) [inline cairo/Context.cs]: :+1:

File: glib/Object.cs (hunk 1 above)
Comment (alanmcgovern) [inline glib/Object.cs]: it'd probably be better to not break it then. There isn't a compelling reason to break this :)

File: glib/Marshaller.cs (hunk 2 above)
Comment (alanmcgovern) [inline glib/Marshaller.cs]: it'd probably be better to not break it then. There isn't a compelling reason to break this :)

Comment (alanmcgovern) [inline glib/Marshaller.cs]: (i'm not sure what this could break)

**Comment by Therzok** [comment]
Sizeof commit got squashed into another commit. Also drastically reduced in size, so won't bother splitting the commit.

**Comment by alanmcgovern** [comment]
Did we undo the API/ABI breakages?

**Comment by Therzok** [comment]
Yes. All of them.

**Pull request merged**